File: spark/src/main/java/systematicchaos/studentmobility/SpecificQuery1.java
/**
 * Student Mobility
 * 
 * Javier Fernández-Bravo Peñuela
 * Universitat Politècnica de València
 * 
 * Specific query 1: How many students, from the same home country as the 22-aged individual,
 * whose home university code is "CIUDA-R" and whose host university code is "S VASTERA",
 * took their Erasmus exchange in the same host university.
 * 
 * spark - SpecificQuery1.java
 */

package systematicchaos.studentmobility;

import java.util.List;
import java.util.ArrayList;

import org.apache.spark.sql.Column;
import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;

import systematicchaos.studentmobility.domain.Student;
import systematicchaos.studentmobility.spark.SparkJob;
import systematicchaos.studentmobility.spark.SparkJobLauncher;
import systematicchaos.studentmobility.spark.SparkTask;
import systematicchaos.studentmobility.util.KeyCountPair;
import systematicchaos.studentmobility.util.StudentBuilder;

public class SpecificQuery1 {
	
	public static class CountryCompanions implements SparkTask<Long> {
		
		private static final String HOME_UNIVERSITY = "E  CIUDA-R";
		private static final String HOST_UNIVERSITY = "S  VASTERA";
		private static final int AGE = 22;
		
		@Override
		public Iterable<KeyCountPair<Long>> computeSparkTask(String[] inputFiles) {
			SparkJob<Student> job = new SparkJob<>(SpecificQuery1.class.getName(),
					Student.class, new StudentBuilder());
			List<KeyCountPair<Long>> companionCount = new ArrayList<>(1);
			
			try {
				Dataset<Student> students = job.readDataset(inputFiles);
				
				Row me = students.filter(students.col("homeInstitution").startsWith(HOME_UNIVERSITY).and(
						students.col("hostInstitution").startsWith(HOST_UNIVERSITY)).and(
								students.col("age").equalTo(AGE)))
						.select(new Column("countryOfHomeInstitution"), new Column("countryOfHostInstitution"))
						.first();
				
				Dataset<Student> companions = students.filter(students.col("countryOfHomeInstitution").equalTo(me.getString(0))
						.and(students.col("countryOfHostInstitution").equalTo(me.getString(1))));
				companionCount.add(new KeyCountPair<>("COUNTRY_COMPANIONS", companions.count()));
			} catch (Exception e) {
				System.err.println(e.getClass().getName() + ": " + e.getMessage());
			} finally {
				job.close();
			}
			
			return companionCount;
		}
	}
	
	// spark-submit --class systematicchaos.studentmobility.SpecificQuery1 --master local[*] target/spark-1.0.0.jar ../data/Student_Mobility.csv [out/spark-1-output]
	public static void main(String[] args) throws Exception {
		SparkJobLauncher.launchSparkJob(new CountryCompanions(), args);
	}
}

File: pandas/specific_query_1.py
#!/usr/bin/python3

"""Specific query 1

Student Mobility

Javier Fernández-Bravo Peñuela
Universitat Politècnica de València

How many students, from the same home country as the 22-aged individual,
whose home university code is "CIUDA-R" and whose host university code
is "S VASTERA", took their Erasmus exchange in the same host university.

pandas - specific_query_1.py
"""

from pandas_job.pandas_task import compute_task
from domain.student import Student

import pandas as pd
from sys import argv

HOME_UNIVERSITY = "CIUDA-R"
HOST_UNIVERSITY = "VASTERA"
AGE = 22

def country_companions(students_df):
    me = students_df.iloc[\
        students_df.index[(students_df['HOMEINSTITUTION'].apply(hi_sub_str).str.startswith(HOME_UNIVERSITY))\
            & (students_df['HOSTINSTITUTION'].apply(hi_sub_str).str.startswith(HOST_UNIVERSITY))\
                & (students_df['AGE'].eq(AGE))][0]]

    companions = students_df[(students_df['COUNTRYOFHOMEINSTITUTION'] == me['COUNTRYOFHOMEINSTITUTION'])\
        & (students_df['COUNTRYOFHOSTINSTITUTION'] == me['COUNTRYOFHOSTINSTITUTION'])]
    return pd.Series({ 'COUNTRY_COMPANIONS': companions.shape[0] })

def hi_sub_str(hi):
    return hi[3:]
    

# python3 specific_query_1.py ../data/Student_Mobility.csv [out/pandas-1-output]
if __name__ == "__main__":
    compute_task(country_companions, Student, argv[1:])
